### PR TITLE
Add device group filter to widgets

### PR DESCRIPTION
**Pull request by Jellyfrog**
Adds a device group filter to most widgets

DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

File: app/Http/Controllers/Table/EventlogController.php [hunk 1]
```diff
@@ -61,7 +62,15 @@ protected function filterFields($request)
      */
     public function baseQuery($request)
     {
-        return Eventlog::hasAccess($request->user())->with('device');
+        $query = Eventlog::hasAccess($request->user())->with('device');
+
+        if ($request->device_group) {
+            $query->whereHas('device.groups', function ($query) use ($request) {
```
Comment (murrant) [inline app/Http/Controllers/Table/EventlogController.php]: Should probably make this a query scope on the Device model (like hasAccess)

Comment (Jellyfrog) [reply app/Http/Controllers/Table/EventlogController.php]: True, will fix.

Comment (murrant) [reply app/Http/Controllers/Table/EventlogController.php]: That would also make it so you don't have to use all those clones below.  Those aren't very nice to look at :)

File: app/Http/Controllers/Table/EventlogController.php [hunk 2]
```diff
@@ -62,13 +62,11 @@ protected function filterFields($request)
...
     public function baseQuery($request)
     {
-        $query = Eventlog::hasAccess($request->user())->with('device');
-
-        if ($request->device_group) {
-            $query->whereHas('device.groups', function ($query) use ($request) {
-                $query->where('id', $request->device_group);
-            });
-        }
+        $query = Eventlog::hasAccess($request->user())->with(['device' => function ($query) use ($request) {
```
Comment (Jellyfrog) [inline app/Http/Controllers/Table/EventlogController.php]: @murrant Like this you had in mind?

**Comment by murrant** [comment]
@Jellyfrog I've merged #9716 
You can use that now, feel free to update the device group query if needed.

**Comment by murrant** [comment]
Hey @Jellyfrog are you planning on finishing this?

**Comment by Jellyfrog** [comment]
> Hey @Jellyfrog are you planning on finishing this?

Hi, yes, forgot about it.

File: app/Models/DeviceRelatedModel.php [hunk 3]
```diff
@@ -36,8 +36,9 @@ public function scopeHasAccess($query, User $user)
 
     public function scopeInDeviceGroup($query, $deviceGroup)
     {
-        $groups = $deviceGroup instanceof DeviceGroup ? $deviceGroup->devices()->pluck('devices.device_id') : [];
-        return $query->whereIn($this->getTable() . '.device_id', $groups);
+        return $query->whereHas('deviceGroup', function ($query) use ($deviceGroup) {
```
Comment (Jellyfrog) [inline app/Models/DeviceRelatedModel.php]: @murrant what do you say about doing like this instead?
Will generate the SQL:
```sql
select * from `eventlog`
where exists (
  select * from `device_group_device`
  where `eventlog`.`device_id` = `device_group_device`.`device_id`
  and `device_group_device`.`device_group_id` = ?
)
``` 
That should be more efficient :)

**Comment by Jellyfrog** [comment]
@murrant 
What do you think about doing like this instead: https://github.com/librenms/librenms/pull/9692/commits/d01d9773d9b358fd28c84019dc9c0f13981d0faa

Will generate the SQL:
```sql
select * from `eventlog`
where exists (
  select * from `device_group_device`
  where `eventlog`.`device_id` = `device_group_device`.`device_id`
  and `device_group_device`.`device_group_id` = ?
)
``` 
That should be more efficient :)

File: app/Models/Port.php [hunk 4]
```diff
@@ -209,13 +209,25 @@ public function scopeHasAccess($query, User $user)
         return $this->hasPortAccess($query, $user);
     }
 
+    public function scopeInDeviceGroup($query, $deviceGroup)
```
Comment (murrant) [inline app/Models/Port.php]: Port is already a DeviceRelatedModel, so you shouldn't need this to be redefined here.

File: app/Models/DeviceRelatedModel.php [hunk 5]
```diff
@@ -46,4 +47,9 @@ public function device()
     {
         return $this->belongsTo('App\Models\Device', 'device_id', 'device_id');
     }
+
+    public function deviceGroup()
+    {
+        return $this->HasMany('App\Models\DeviceGroupDevice', 'device_id', 'device_id');
```
Comment (murrant) [inline app/Models/DeviceRelatedModel.php]: You should copy the groups() definition from the device model.  No need to create a model for the pivot table.

Then you just need to make everything consistent.

Comment (Jellyfrog) [reply app/Models/DeviceRelatedModel.php]: I know it's not needed, but it optimizes the SQL.
Current:
```sql
SELECT *
FROM `eventlog`
WHERE EXISTS
    (SELECT *
     FROM `device_group_device`
     WHERE `eventlog`.`device_id` = `device_group_device`.`device_id`
       AND `device_group_device`.`device_group_id` = '2')
ORDER BY `datetime` DESC
LIMIT 50
OFFSET 0;
```

With the following change:
```php
return $query->whereHas('groups', function ($query) use ($deviceGroup) {
    $query->where('device_groups.id', $deviceGroup);
});
```
The SQL becomes:
```sql
SELECT *
FROM `eventlog`
WHERE EXISTS
    (SELECT *
     FROM `device_groups`
     INNER JOIN `device_group_device` ON `device_groups`.`id` = `device_group_device`.`device_group_id`
     WHERE `eventlog`.`event_id` = `device_group_device`.`device_id`
       AND `device_groups`.`id` = '2')
ORDER BY `datetime` DESC
LIMIT 50
OFFSET 0
```
So this is simply a optimized version.

File: app/Models/Device.php [hunk 6]
```diff
@@ -15,7 +15,7 @@
 use LibreNMS\Util\Url;
 use LibreNMS\Util\Time;
 
-class Device extends BaseModel
+class Device extends DeviceRelatedModel
```
Comment (Jellyfrog) [inline app/Models/Device.php]: @murrant This is up to you, a bit quirky, but it saves having to duplicate functions from `DeviceRelatedModel.php`

Comment (murrant) [reply app/Models/Device.php]: I think I prefer the duplicated function...  It doesn't contain any code so the duplication is minor.

File: app/Models/DeviceRelatedModel.php (hunk 5 above)
Comment (murrant) [reply app/Models/DeviceRelatedModel.php]: This returns a DeviceGroupDevice model not a DeviceGroup model.

File: resources/views/widgets/settings/availability-map.blade.php [hunk 7]
```diff
@@ -56,17 +56,17 @@
 
     <div class="form-group">
         <label for="device_group-{{ $id }}" class="control-label">@lang('Device group')</label>
-        <select class="form-control" name="device_group" id="device_group-{{ $id }}"  data-placeholder="@lang('All Devices')">
+        <select class="form-control" name="device_group" id="device_group-{{ $id }}" data-placeholder="@lang('All Device groups')">
```
Comment (murrant) [inline resources/views/widgets/settings/availability-map.blade.php]: This should still be "All Devices".

File: resources/views/widgets/settings/availability-map.blade.php [hunk 8]
```diff
@@ -56,17 +56,17 @@
...
+                <option value="{{ $device_group->id }}" selected>{{ $device_group->name }}</option>
             @endif
         </select>
     </div>
 @endsection
 
 @section('javascript')
     <script type="text/javascript">
-        init_select2('#device_group-{{ $id }}', 'device-group', {}, '{{ $device_group ? $device_group->id : '' }}');
+        init_select2('#device_group-{{ $id }}', 'device-group', {});
```
Comment (murrant) [inline resources/views/widgets/settings/availability-map.blade.php]: When you remove this, is the current device group pre-selected?

Comment (Jellyfrog) [reply resources/views/widgets/settings/availability-map.blade.php]: Yes, and now it shows the name from the `<option>' instead of the id

Comment (murrant) [reply resources/views/widgets/settings/availability-map.blade.php]: Oh yeah, you have to pass an object there with name and id (then you can avoid the select in html).  I'm not sure which one is better...

File: app/Models/Device.php [hunk 9]
```diff
@@ -476,6 +483,11 @@ public function groups()
         return $this->belongsToMany('App\Models\DeviceGroup', 'device_group_device', 'device_id', 'device_group_id');
     }
 
+    public function groupDevices()
```
Comment (murrant) [inline app/Models/Device.php]: I don't want this relationship or the DeviceGroupDevice model.

File: app/Models/DeviceGroupDevice.php [hunk 10]
```diff
@@ -0,0 +1,10 @@
+<?php
```
Comment (murrant) [inline app/Models/DeviceGroupDevice.php]: Remove this.

File: app/Models/DeviceRelatedModel.php [hunk 11]
```diff
@@ -46,4 +47,9 @@ public function device()
     {
         return $this->belongsTo('App\Models\Device', 'device_id', 'device_id');
     }
+
+    public function groupDevices()
```
Comment (murrant) [inline app/Models/DeviceRelatedModel.php]: remove

**Comment by murrant** [comment]
You should either just check the table directly in your scope:
```
return $query->whereIn('device_id', function ($query) use ($deviceGroup) {
    /** @var Builder $query */
    $query->select('device_id')->whereIn('device_id', function ($query) use ($deviceGroup) {
    /** @var Builder $query */
    $query->select('device_id')
        ->from('device_group_device')
        ->where('device_group_id', $deviceGroup);
})
        ->from('device_group_device')
        ->where('device_group_id', $deviceGroup);
})
```

or change the basequery so instead of `Device::query()` (query is often ommited)

`DeviceGroup::find($deviceGroup)->devices()`

or a fake device group if you want to avoid the lookup
```
$dg = new DeviceGroup();
$dg->id = $deviceGroup;
$query = $dg->devices()->hasAccess...
```

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/v1-55-release-changelog-august-2019/9428/1

**Pull request merged**